### PR TITLE
*: add builtin func interface.

### PR DESCRIPTION
**Pull request by hanfei1991**
prepare to refactor function model.
@shenli @zimulala @coocood @tiancaiamao PTAL

### CODE REVIEW

**Comment by shenli** [comment]
Please add some comments for the new struct/interface.

File: expression/builtin.go [hunk 1]
```diff
@@ -20,12 +20,89 @@ package expression
...
 	"strings"
 
+	"github.com/juju/errors"
 	"github.com/pingcap/tidb/ast"
 	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/parser/opcode"
 	"github.com/pingcap/tidb/util/types"
 )
 
+type baseBuiltinFunc struct {
```
Comment (shenli) [inline expression/builtin.go]: var _  builtinFunc = &baseBuiltinFunc{}

Comment (hanfei1991) [inline expression/builtin.go]: baseBuiltinFunc doesn't implement builtinFunc

File: expression/builtin.go [hunk 2]
```diff
@@ -20,12 +20,90 @@ package expression
...
+}
+
+// builtinFunc stands for a particular function signature.
+type builtinFunc interface {
+	// eval does evaluation by the given row.
+	eval([]types.Datum) (types.Datum, error)
+	// getArgs returns the arguments expressions.
+	getArgs() []Expression
+	// isDeterministic checks if a function is deterministic.
+	// a function is deterministic if it returns different results for same inputs.
```
Comment (coocood) [inline expression/builtin.go]: Do you mean `returns same results for same inputs`?
Please give a couple of non-deterministic functions.

Comment (hanfei1991) [inline expression/builtin.go]: random returns different results for same input

File: expression/builtin.go [hunk 3]
```diff
@@ -20,12 +20,91 @@ package expression
...
+	for i, arg := range b.args {
+		b.argValues[i], err = arg.Eval(row, b.ctx)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return b.argValues, nil
+}
+
+func (b *baseBuiltinFunc) isDeterministic() bool {
```
Comment (coocood) [inline expression/builtin.go]: This can leave for real builtinFunc to implement.

**Comment by shenli** [comment]
LGTM

**Pull request merged**